subsys/net/lib/nrf_cloud/src/nrf_cloud_cell_pos.c

```diff
@@ -0,0 +1,312 @@
+		modem_info->network.area_code.value) ||
+	    !cJSON_AddNumberToObject(data_obj, CELL_POS_JSON_CELL_ID_KEY,
+		(uint32_t)modem_info->network.cellid_dec) ||
+	    !cJSON_AddNumberToObject(data_obj, CELL_POS_JSON_PHYCID_KEY, 0)) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int json_add_modem_info(cJSON *const data_obj)
+{
+	__ASSERT_NO_MSG(data_obj != NULL);
+
+	struct modem_param_info modem_info = {0};
+	int err;
+
+	err = get_modem_info(&modem_info);
+	if (err) {
+		return err;
+	}
+
+	return json_format_data_obj(data_obj, &modem_info);
+}
+
+static int json_send_to_cloud(cJSON *const cell_pos_request)
+{
+	__ASSERT_NO_MSG(cell_pos_request != NULL);
+
+	char *msg_string;
+	int err;
+
+	msg_string = cJSON_PrintUnformatted(cell_pos_request);
+	if (!msg_string) {
+		LOG_ERR("Could not allocate memory for Cell Pos request message");
+		return -ENOMEM;
+	}
+
+	LOG_DBG("Created Cell Pos request: %s", log_strdup(msg_string));
+
+	struct nct_dc_data msg = {
+		.data.ptr = msg_string,
+		.data.len = strlen(msg_string)
+	};
+
+	err = nct_dc_send(&msg);
+	if (err) {
+		LOG_ERR("Failed to send Cell Pos request, error: %d", err);
+	} else {
+		LOG_DBG("Cell Pos request sent");
+	}
+
+	k_free(msg_string);
+
+	return err;
+}
+
+bool json_item_string_exists(const cJSON *const obj, const char *const key, const char *const val)
+{
+	__ASSERT_NO_MSG(obj != NULL);
+	__ASSERT_NO_MSG(key != NULL);
+
+	char *str_val;
+	cJSON *item = cJSON_GetObjectItem(obj, key);
+
+	if (!item) {
+		return false;
+	}
+
+	if (!val) {
+		return cJSON_IsNull(item);
+	}
+
+	str_val = cJSON_GetStringValue(item);
+	if (!str_val) {
+		return false;
+	}
+
+	return (strcmp(str_val, val) == 0);
+}
+
+static int json_parse_cell_location(const cJSON *const cell_pos_obj,
+				    struct nrf_cloud_cell_pos_result *result)
+{
+	__ASSERT_NO_MSG(cell_pos_obj);
+
+	cJSON *lat, *lon, *unc;
+
+	lat = cJSON_GetObjectItem(cell_pos_obj, CELL_POS_JSON_CELL_LOC_KEY_LAT);
+	lon = cJSON_GetObjectItem(cell_pos_obj, CELL_POS_JSON_CELL_LOC_KEY_LON);
+	unc = cJSON_GetObjectItem(cell_pos_obj, CELL_POS_JSON_CELL_LOC_KEY_UNCERT);
+
+	if (!cJSON_IsNumber(lat) || !cJSON_IsNumber(lon) || !cJSON_IsNumber(unc)) {
+		LOG_DBG("Expected items not found in cell-based location msg");
+		return 1; /* 1: cell-based location not found */
+	}
+
+	result->lat = lat->valuedouble;
+	result->lon = lon->valuedouble;
+	result->unc = unc->valueint;
+
+	LOG_DBG("Cell location: (%lf, %lf), unc: %d, type: %d",
+		result->lat, result->lon, result->unc, result->type);
+
+	return 0;  /* 0: cell-based location found */
+}
+
+static int parse_cell_location_response(const char *const buf,
+					struct nrf_cloud_cell_pos_result *result)
+{
+	int ret;
+	cJSON *cell_pos_obj;
+	cJSON *data_obj;
+
+	if (buf == NULL) {
+		return -EINVAL;
+	}
+
+	cell_pos_obj = cJSON_Parse(buf);
+	if (!cell_pos_obj) {
+		LOG_DBG("No JSON found for cell location");
+		return 1; /* 1: cell-based location not found */
+	}
+
+	ret = -EBADMSG;
```

Comment:
`ret = -EBADMSG;` should be moved *after* the JSON checks for `CELL_POS_JSON_MSG_TYPE_KEY` and `CELL_POS_JSON_APPID_KEY`

it is only an error (negative return) if the JSON message contains a cell loc message (appID == SCELL or MCELL and messageType == DATA) and there is a problem parsing the data payload.  
i believe the original code was incorrect :(

Reply:
It will be initialized to be 1, and return 1 if  JSON checks for CELL_POS_JSON_MSG_TYPE_KEY and CELL_POS_JSON_APPID_KEY fails.

Reply:
Yes, cloud data comes in to the application through a generic event `CLOUD_EVT_DATA_RECEIVED`.
The data payload could be JSON or it could be binary.

Reply:
Keep returning 1 as if not JSON then not reply of SCELL request